pkg/receive/handler.go

```diff
@@ -592,43 +582,102 @@ func (h *Handler) RemoteWrite(ctx context.Context, r *storepb.WriteRequest) (*st
 	}
 }
 
-// countCause counts the number of errors within the given error
-// whose causes satisfy the given function.
-// countCause will inspect the error's cause or, if the error is a MultiError,
-// the cause of each contained error but will not traverse any deeper.
-func countCause(err error, f func(error) bool) int {
-	errs, ok := err.(terrors.MultiError)
-	if !ok {
-		errs = []error{err}
-	}
-	var n int
-	for i := range errs {
-		if f(errors.Cause(errs[i])) {
-			n++
-		}
-	}
-	return n
-}
-
 // isConflict returns whether or not the given error represents a conflict.
 func isConflict(err error) bool {
 	if err == nil {
 		return false
 	}
-	return err == conflictErr ||
+	return err == errConflict ||
 		err == storage.ErrDuplicateSampleForTimestamp ||
 		err == storage.ErrOutOfOrderSample ||
 		err == storage.ErrOutOfBounds ||
-		err.Error() == strconv.Itoa(http.StatusConflict) ||
 		status.Code(err) == codes.AlreadyExists
 }
 
 // isNotReady returns whether or not the given error represents a not ready error.
 func isNotReady(err error) bool {
-	return err == tsdb.ErrNotReady ||
+	return err == errNotReady ||
+		err == tsdb.ErrNotReady ||
 		status.Code(err) == codes.Unavailable
 }
 
+// countCause counts the number of errors within the given error slice
+// whose causes satisfy the given function.
+// cause will inspect the error's cause or, if the error is a MultiError,
+// the cause of each contained error but will not traverse any deeper.
```

Comment:
This comment is no longer accurate: this function _does_ traverse deeply, in fact it can recurse infinitely, which is new functionality compared to before.